test/functional/sqlsrv/sqlsrv_ae_datetimes_as_strings.phpt

```diff
@@ -0,0 +1,595 @@
+--TEST--
+Test various date and time types with AE and ReturnDatesAsStrings set to true
+--SKIPIF--
+<?php require('skipif_versions_old.inc'); ?>
+--FILE--
+<?php
+// Check for expected errors. These are expected in cases where the dates and
+// times do not comply with ODBC standards.
+function ExecutePreparedStmt($stmt)
+{
+    if ($stmt == false) {
+        fatalError("Failure in sqlsrv_prepare");
+    } else {
+        $r = sqlsrv_execute($stmt);
+        if ($r == false) {
+            $errors = sqlsrv_errors();
+            if ($errors[0]['SQLSTATE'] != '22018' and
+                $errors[0]['SQLSTATE'] != '22008' and
+                $errors[0]['SQLSTATE'] != '22007' and
+                $errors[0]['SQLSTATE'] != '07006') {
+                print_r($errors);
+                fatalError("Unexpected error");
+            }
+        }
+    }
+}
+
+// Compare dates retrieved from the database with the date used for testing.
+// $testingDate is an array of strings like '2002-01-31 23:59:59.049876' and
+// $retrieved_date is the string date of the format 'Y-m-d H:i:s.u', which 
+// the format returned by the date_format() calls below or by retrieval as 
+// strings, unless $datetimetype is date or time. In that case, if 
+// $returnStrings is false and $datetimetype is 'date', the time defaults to
+// 00:00:00.0000, and if $datetimetype is 'time' and $returnStrings is false,
+// the date defaults to the current date. If however $returnStrings is true
+// and $datetimetype is 'date', $retrieved_date is only a date,
+// and if $datetimetype is 'time' and $returnStrings is true, $retrieved_date
+// only a time.
+// The concatenations involving zeroes below are to make direct string
+// comparisons feasible. Also, because PHP maxes out at microsecond precision
+// and SQL Server maxes out at 0.1 microsecond precision, the more precise 
+// types require an extra 0 for some comparisons.
+function CompareDateTime($datetimetype, $returnStrings, &$testingDate, $retrievedDate)
+{
+    $test_date_time = array();
+
+    for ($i=0; $i<sizeof($testingDate); ++$i) {
+        $test_date_time[] = explode(" ", $testingDate[$i]);
+    }
+    
+    $ret_date_time = explode(" ", $retrievedDate);
+    if (sizeof($ret_date_time)>1) {
+        $datetimeonly = $ret_date_time[0]." ".$ret_date_time[1];
+    }
+    if (sizeof($ret_date_time)>2) {
+        $timezone = $ret_date_time[2];
+        $dtoffset = $ret_date_time[0]." ".substr($ret_date_time[1], 0, -1)." ".$ret_date_time[2];
+    }
+   
+    if ($returnStrings == true) {
+        switch ($datetimetype) {
```

Comment:
Suggestion: one comparison for datetime values as strings and one method for datetime objects?

Also, the variable names are too similar and hard to understand the logic. Is it possible to distinguish them by using adjectives like "actual" and "expected"?

Reply:
Everything is compared as strings, because when a DateTime object is returned it is first formatted as a string using date_format() before being passed to this function.